src/server/ua_subscription_datachange.c

```diff
@@ -146,10 +148,108 @@ void MonitoredItem_ensureQueueSpace(UA_MonitoredItem *mon) {
+            if ((((const UA_UInt16*)data1)[index] - ((const UA_UInt16*)data2)[index]) <= deadbandValue)
+                return false;
+        } else {
+            if ((((const UA_UInt16*)data2)[index] - ((const UA_UInt16*)data1)[index]) <= deadbandValue)
+                return false;
+        }
+    } else
+    if (type == &UA_TYPES[UA_TYPES_INT32]) {
+        if (abs(((const UA_Int32*)data1)[index] - ((const UA_Int32*)data2)[index]) <= deadbandValue)
+            return false;
+    } else
+    if (type == &UA_TYPES[UA_TYPES_UINT32]) {
+        if (((const UA_UInt32*)data1)[index] > ((const UA_UInt32*)data2)[index]) {
+            if ((((const UA_UInt32*)data1)[index] - ((const UA_UInt32*)data2)[index]) <= deadbandValue)
+                return false;
+        } else {
+            if ((((const UA_UInt32*)data2)[index] - ((const UA_UInt32*)data1)[index]) <= deadbandValue)
+                return false;
+        }
+    } else
+    if (type == &UA_TYPES[UA_TYPES_INT64]) {
+        if (llabs(((const UA_Int64*)data1)[index] - ((const UA_Int64*)data2)[index]) <= deadbandValue)
+            return false;
+    } else
+    if (type == &UA_TYPES[UA_TYPES_UINT64]) {
+        if (((const UA_UInt64*)data1)[index] > ((const UA_UInt64*)data2)[index]) {
+            if ((((const UA_UInt64*)data1)[index] - ((const UA_UInt64*)data2)[index]) <= deadbandValue)
+                return false;
+        } else {
+            if ((((const UA_UInt64*)data2)[index] - ((const UA_UInt64*)data1)[index]) <= deadbandValue)
+                return false;
+        }
+    } else
+    if (type == &UA_TYPES[UA_TYPES_FLOAT]) {
+        if (fabsf(((const UA_Float*)data1)[index] - ((const UA_Float*)data2)[index]) <= deadbandValue)
+            return false;
+    } else
+    if (type == &UA_TYPES[UA_TYPES_DOUBLE]) {
+        if (fabs(((const UA_Double*)data1)[index] - ((const UA_Double*)data2)[index]) <= deadbandValue)
+            return false;
+    }
+    return true;
+}
+
+static UA_INLINE UA_Boolean
+updateNeededForFilteredValue(const UA_Variant *value, const UA_Variant *oldValue, const UA_Double deadbandValue) {
+    if (value->arrayLength != oldValue->arrayLength) {
+        return true;
+    }
+    if (value->type != oldValue->type) {
+        return true;
+    }
+    if (UA_Variant_isScalar(value)) {
+        return outOfDeadBand(value->data, oldValue->data, 0, value->type, deadbandValue);
+    } else {
+        for (size_t i = 0; i < value->arrayLength; ++i) {
+            if (outOfDeadBand(value->data, oldValue->data, i, value->type, deadbandValue))
+                return true;
+        }
+    }
+    return false;
+}
+
 /* Errors are returned as no change detected */
 static UA_Boolean
 detectValueChangeWithFilter(UA_MonitoredItem *mon, UA_DataValue *value,
                             UA_ByteString *encoding) {
+    if (isDataTypeNumeric(value->value.type)
+            && (mon->filter.trigger == UA_DATACHANGETRIGGER_STATUSVALUE
```

Comment:
`mon->filter` can also be NULL? So there should be a NULL check?

Reply:
filter is a value based member and can't be null. It is no pointer in the monitored item.